modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
+            // the value. Also before returning value, puts the value in cache.
+            BiFunction<Tuple<ICacheKey<K>, V>, Throwable, Void> handler = (pair, ex) -> {
+                if (pair != null) {
+                    try (ReleasableLock ignore = writeLock.acquire()) {
+                        onHeapCache.put(pair.v1(), pair.v2());
+                    } catch (Exception e) {
+                        // TODO: Catch specific exceptions to know whether this resulted from cache or underlying removal
+                        // listeners/stats. Needs better exception handling at underlying layers.For now swallowing
+                        // exception.
+                        logger.warn("Exception occurred while putting item onto heap cache", e);
+                    }
+                } else {
+                    if (ex != null) {
+                        logger.warn("Exception occurred while trying to compute the value", ex);
+                    }
+                }
+                completableFutureMap.remove(key);// Remove key from map as not needed anymore.
+                return null;
+            };
+            V value = null;
+            if (future == null) {
+                future = completableFutureMap.get(key);
+                future.handle(handler);
+                try {
+                    value = loader.load(key);
+                } catch (Exception ex) {
+                    future.completeExceptionally(ex);
+                    throw new ExecutionException(ex);
+                }
+                if (value == null) {
+                    NullPointerException npe = new NullPointerException("Loader returned a null value");
+                    future.completeExceptionally(npe);
+                    throw new ExecutionException(npe);
+                } else {
+                    future.complete(new Tuple<>(key, value));
+                }
+            } else {
+                try {
+                    value = future.get().v2();
+                } catch (InterruptedException ex) {
+                    throw new IllegalStateException(ex);
+                }
             }
             return value;
-        } else {
-            // Handle stats for an initial hit from getValueFromTieredCache()
-            if (cacheValueTuple.v2().equals(TIER_DIMENSION_VALUE_ON_HEAP)) {
-                // A hit for the heap tier
-                statsHolder.incrementHits(heapDimensionValues);
-            } else if (cacheValueTuple.v2().equals(TIER_DIMENSION_VALUE_DISK)) {
-                // Miss for the heap tier, hit for the disk tier
-                statsHolder.incrementMisses(heapDimensionValues);
-                statsHolder.incrementHits(diskDimensionValues);
+        }
+
+        @Override
+        public void invalidate(ICacheKey<K> key) {
+            for (Map.Entry<ICache<K, V>, TierInfo> cacheEntry : caches.entrySet()) {
```

Comment:
Does the below comment no longer apply to invalidate logic?

```
// We are trying to invalidate the key from all caches though it would be present in only of them.
        // Doing this as we don't know where it is located. We could do a get from both and check that, but what will
        // also trigger a hit/miss listener event, so ignoring it for now.
        // We don't update stats here, as this is handled by the removal listeners for the tiers.
```

Reply:
Yeah this comment was mostly stale. It was added initially when we were not maintaining cache stats explicitly within TieredCache and using listeners for that.
Though some parts still hold true ie why are we invalidating from both/all caches and the reason being for simplicity.    So removed it altogether. Will change this logic separately if needed.